sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotCheckZeroSizeCollection.cs

```diff
@@ -0,0 +1,144 @@
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+using System;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class DoNotCheckZeroSizeCollection : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S3981";
+        private const string MessageFormat = "The {0} of '{1}' is always '>=0', so fix this test to get the real expected behavior.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+                {
+                    var gteNode = (BinaryExpressionSyntax)c.Node;
+                    CheckCondition(c, gteNode.Left, gteNode.Right);
+                }, SyntaxKind.GreaterThanOrEqualExpression);
+
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+                {
+                    var lteNode = (BinaryExpressionSyntax)c.Node;
+                    CheckCondition(c, lteNode.Right, lteNode.Left);
+                }, SyntaxKind.LessThanOrEqualExpression);
+        }
+
+        private static void CheckCondition(SyntaxNodeAnalysisContext context, ExpressionSyntax expressionValueNode,
+            ExpressionSyntax constValueNode)
+        {
+            if (!IsConstantZero(context, constValueNode))
+            {
+                return;
+            }
+
+            ISymbol symbol = GetSymbol(context, expressionValueNode);
+
+            string symbolType = null;
+            if (IsArrayLengthProperty(symbol))
+            {
+                symbolType = "Array";
+            }
+            else if (IsEnumerableCountMethod(symbol))
+            {
+                symbolType = "IEnumerable<T>";
+            }
+            else if (IsCollectionProperty(symbol))
+            {
+                symbolType = "ICollection";
+            }
+
+            if (symbolType == null)
+            {
+                return;
+            }
+
+            context.ReportDiagnostic(Diagnostic.Create(rule, context.Node.GetLocation(),
+                symbol.Name.ToLower(), symbolType));
+        }
+
+        private static ISymbol GetSymbol(SyntaxNodeAnalysisContext context, ExpressionSyntax expressionValueNode)
+        {
+            while (true)
```

Comment:
Arf IMO `while (true)` should be avoided as much as possible. In this case it is easy to move the exit condition in the while.